File: claudespaces/container.py
import os
import subprocess
from pathlib import Path

import docker
from docker.types import Mount

from claudespaces.host_config import DEFAULT_PORT, SHIMS_PATH

CONTAINER_USER = "root"
_CLAUDESPACES_HOST_SRC = Path(__file__).parent / "support" / "bin" / "claudespaces-host"


def get_running_container_ids(docker_client) -> set[str]:
    containers = docker_client.containers.list(filters={"status": "running"})
    return {c.id for c in containers}


def create_container(docker_client, image: str, dirs: list[str], state_dir: Path, host_port: int = DEFAULT_PORT, additional_mounts: list[dict] | None = None) -> str:
    basenames = [os.path.basename(d) for d in dirs]
    if len(basenames) != len(set(basenames)):
        dup = next(b for b in basenames if basenames.count(b) > 1)
        raise ValueError(f"Directories share the same basename: {dup!r}")

    mounts = []

    for d in dirs:
        mounts.append(Mount(
            target=f"/workspace/{os.path.basename(d)}",
            source=d,
            type="bind",
            read_only=False,
        ))

    # Per-workspace state mounts (always added; cli.py ensures these exist)
    mounts.append(Mount(
        target="/root/.claude.json",
        source=str(state_dir / "claude.json"),
        type="bind",
        read_only=False,
    ))
    mounts.append(Mount(
        target="/root/.claude/projects",
        source=str(state_dir / "projects"),
        type="bind",
        read_only=False,
    ))

    # Bind-mount the entrypoint from the installed package so it's always current,
    # regardless of which image version is cached.
    entrypoint_src = Path(__file__).parent / "support" / "bin" / "entrypoint.sh"
    if entrypoint_src.exists():
        mounts.append(Mount(
            target="/claudespaces/entrypoint.sh",
            source=str(entrypoint_src),
            type="bind",
            read_only=True,
        ))

    # Selective host config mounts (conditional on existence)
    home = Path.home()
    host_mounts = [
        (home / ".claude" / "settings.json", "/claudespaces/host/settings.json"),
        (home / ".claude" / "plugins", "/claudespaces/host/plugins"),
        (home / ".claude" / ".credentials.json", "/claudespaces/host/credentials.json"),
    ]
    for source, target in host_mounts:
        if source.exists():
            mounts.append(Mount(
                target=target,
                source=str(source),
                type="bind",
                read_only=True,
            ))

    if SHIMS_PATH.exists():
        mounts.append(Mount(
            target="/claudespaces/shims.json",
            source=str(SHIMS_PATH),
            type="bind",
            read_only=True,
        ))

    if _CLAUDESPACES_HOST_SRC.exists():
        mounts.append(Mount(
            target="/claudespaces/bin/claudespaces-host",
            source=str(_CLAUDESPACES_HOST_SRC),
            type="bind",
            read_only=True,
        ))

    for m in (additional_mounts or []):
        mounts.append(Mount(
            target=m["target"],
            source=m["source"],
            type="bind",
            read_only=m["read_only"],
        ))

    container = docker_client.containers.create(
        image=image,
        tty=True,
        stdin_open=True,
        user=CONTAINER_USER,
        working_dir="/workspace",
        mounts=mounts,
        extra_hosts={"host.docker.internal": "host-gateway"},
        environment={
            "IS_SANDBOX": "1",
            "HOST_HOME": str(Path.home()),
            "CLAUDESPACES_HOST_PORT": str(host_port),
        },
    )
    return container.id


def attach_container(container_id: str) -> None:
    # Start the container (runs entrypoint: config setup + sleep infinity).
    # Then exec with -it so Docker puts the host terminal into raw mode —
    # docker start -ai skips that step, breaking Enter in TUI applications.
    subprocess.run(["docker", "start", container_id], check=True, capture_output=True)
    cmd = ["docker", "exec", "-it"]
    for var in ("TERM", "COLORTERM", "PS1"):
        val = os.environ.get(var)
        if val:
            cmd += ["-e", f"{var}={val}"]
    cmd += [container_id, "/claudespaces/entrypoint.sh"]
    subprocess.run(cmd)


def stop_container(docker_client, container_id: str) -> None:
    docker_client.containers.get(container_id).stop()


def remove_container(docker_client, container_id: str) -> None:
    try:
        docker_client.containers.get(container_id).remove(force=True)
    except docker.errors.NotFound:
        pass

File: src/Claudespaces/Container.hs

module Claudespaces.Container
  ( checkBasenameCollision
  , buildMounts
  , hostClaudePaths
  , resolveHostMounts
  , buildEnv
  , mountToArgs
  , createContainer
  , attachContainer
  , getRunningContainerIds
  , stopContainer
  , removeContainer
  ) where

import           Control.Monad          (void)
import           System.Directory       (doesFileExist, doesDirectoryExist)
import           Data.List              (group, sort)
import qualified Data.Set               as Set
import           Data.Set               (Set)
import           Data.Text              (Text)
import qualified Data.Text              as T
import           System.Exit            (ExitCode (..))
import           System.FilePath        (takeBaseName, (</>))
import           System.Process         ( createProcess
                                        , proc
                                        , readProcess
                                        , StdStream (..)
                                        , std_in
                                        , std_out
                                        , std_err
                                        , delegate_ctlc
                                        , waitForProcess
                                        )

import           Claudespaces.Config    (Mount (..))
import           Claudespaces.Error     (AppError (..))

-- ---------------------------------------------------------------------------
-- Pure functions
-- ---------------------------------------------------------------------------

checkBasenameCollision :: [FilePath] -> Either AppError ()
checkBasenameCollision dirs =
  let basenames = map takeBaseName dirs
      dups = map head . filter ((> 1) . length) . group . sort $ basenames
  in case dups of
    []    -> pure ()
    (d:_) -> Left (BasenameCollision d)

buildMounts :: [FilePath] -> FilePath -> Int -> [Mount] -> FilePath -> [Mount]
buildMounts dirs stateDir _hostPort additionalMounts homePath =
  userMounts ++ stateMounts ++ hostMounts ++ additionalMounts
  where
    userMounts =
      [ Mount
          { source   = T.pack d
          , target   = T.pack $ "/workspace/" <> takeBaseName d
          , readOnly = False
          }
      | d <- dirs
      ]

    stateMounts =
      [ Mount
          { source   = T.pack (stateDir </> "claude.json")
          , target   = "/root/.claude.json"
          , readOnly = False
          }
      , Mount
          { source   = T.pack (stateDir </> "projects")
          , target   = "/root/.claude/projects"
          , readOnly = False
          }
      ]

    hostMounts =
      [ Mount
          { source   = T.pack (homePath </> ".claudespaces" </> "shims.json")
          , target   = "/claudespaces/shims.json"
          , readOnly = True
          }
      ]

-- | Host paths to mount read-only into the container if they exist on the host.
hostClaudePaths :: FilePath -> [(FilePath, String)]
hostClaudePaths homePath =
  [ (homePath </> ".claude" </> "settings.json", "/claudespaces/host/settings.json")
  , (homePath </> ".claude" </> "plugins",       "/claudespaces/host/plugins")
  , (homePath </> ".claude" </> "credentials.json", "/claudespaces/host/credentials.json")
  ]

resolveHostMounts :: FilePath -> IO [Mount]
resolveHostMounts homePath = do
  let paths = hostClaudePaths homePath
  fmap concat $ mapM checkAndMount paths
  where
    checkAndMount (src, tgt) = do
      fileExists <- doesFileExist src
      dirExists  <- doesDirectoryExist src
      if fileExists || dirExists
        then pure [Mount (T.pack src) (T.pack tgt) True]
        else pure []

buildEnv :: Int -> FilePath -> [(String, String)]
buildEnv hostPort homePath =
  [ ("IS_SANDBOX", "1")
  , ("HOST_HOME", homePath)
  , ("CLAUDESPACES_HOST_PORT", show hostPort)
  ]

mountToArgs :: Mount -> [String]
mountToArgs m =
  [ "--mount"
  , "type=bind,source=" ++ T.unpack m.source
      ++ ",target=" ++ T.unpack m.target
      ++ if m.readOnly then ",readonly" else ""
  ]

-- ---------------------------------------------------------------------------
-- IO wrappers (not tested — thin shell-outs)
-- ---------------------------------------------------------------------------

createContainer :: Text -> [Mount] -> [(String, String)] -> IO Text
createContainer image mounts envVars = do
  let mountArgs = concatMap mountToArgs mounts
      envArgs   = concatMap (\(k, v) -> ["-e", k <> "=" <> v]) envVars
      args      = [ "create", "--tty", "--interactive"
                  , "--user", "root"
                  , "-w", "/workspace"
                  , "--add-host", "host.docker.internal:host-gateway"
                  ]
                  ++ mountArgs
                  ++ envArgs
                  ++ [T.unpack image]
  out <- readProcess "docker" args ""
  pure . T.strip . T.pack $ out

attachContainer :: Text -> IO ()
attachContainer containerId = do
  let cid = T.unpack containerId
  void $ readProcess "docker" ["start", cid] ""
  (_, _, _, ph) <- createProcess
    (proc "docker"
      [ "exec", "-it"
      , "-e", "TERM=xterm-256color"
      , cid
      , "/claudespaces/entrypoint.sh"
      ])
    { std_in  = Inherit
    , std_out = Inherit
    , std_err = Inherit
    , delegate_ctlc = True
    }
  void $ waitForProcess ph

getRunningContainerIds :: IO (Set Text)
getRunningContainerIds = do
  out <- readProcess "docker" ["ps", "-q", "--no-trunc", "--filter", "status=running"] ""
  let ids = filter (not . T.null) . map T.strip . T.lines . T.pack $ out
  pure $ Set.fromList ids

stopContainer :: Text -> IO ()
stopContainer containerId =
  void $ readProcess "docker" ["stop", T.unpack containerId] ""

removeContainer :: Text -> IO ()
removeContainer containerId = do
  (_, _, _, ph) <- createProcess
    (proc "docker" ["rm", "-f", T.unpack containerId])
    { std_in  = Inherit
    , std_out = Inherit
    , std_err = Inherit
    }
  void $ waitForProcess ph
